packages/flutter/test/material/carousel_test.dart

```diff
@@ -0,0 +1,461 @@
+            reverse: true,
+            padding: EdgeInsets.zero,
+            children: List<Widget>.generate(10, (int index) {
+              return Center(
+                child: Text('Item $index'),
+              );
+            }),
+          ),
+        ),
+      )
+    );
+    await tester.pumpAndSettle();
+
+    expect(getItem(0), findsOneWidget);
+    final Rect rect0 = tester.getRect(getItem(0));
+    // Item 0 should be placed on the end of the screen.
+    expect(rect0, const Rect.fromLTRB(600.0, 0.0, 800.0, 600.0));
+
+    expect(getItem(1), findsOneWidget);
+    final Rect rect1 = tester.getRect(getItem(1));
+    // Item 1 should be placed before item 0.
+    expect(rect1, const Rect.fromLTRB(400.0, 0.0, 600.0, 600.0));
+
+    expect(getItem(2), findsOneWidget);
+    final Rect rect2 = tester.getRect(getItem(2));
+    // Item 2 should be placed before item 1.
+    expect(rect2, const Rect.fromLTRB(200.0, 0.0, 400.0, 600.0));
+
+    expect(getItem(3), findsOneWidget);
+    final Rect rect3 = tester.getRect(getItem(3));
+    // Item 3 should be placed before item 2.
+    expect(rect3, const Rect.fromLTRB(0.0, 0.0, 200.0, 600.0));
+  });
+
+  testWidgets('Carousel respects shrinkExtent', (WidgetTester tester) async {
+    await tester.pumpWidget(
+      MaterialApp(
+        home: Scaffold(
+          body: Carousel(
+            itemExtent: 350,
+            shrinkExtent: 300,
+            children: List<Widget>.generate(10, (int index) {
+              return Center(
+                child: Text('Item $index'),
+              );
+            }),
+          ),
+        ),
+      )
+    );
+    await tester.pumpAndSettle();
+
+    final Rect rect0 = tester.getRect(getItem(0));
+    expect(rect0, const Rect.fromLTRB(0.0, 0.0, 350.0, 600.0));
+
+    final Rect rect1 = tester.getRect(getItem(1));
+    expect(rect1, const Rect.fromLTRB(350.0, 0.0, 700.0, 600.0));
+
+    final Rect rect2 = tester.getRect(getItem(2));
+    // The extent of item 2 is 300, and only 100 is on screen.
+    expect(rect2, const Rect.fromLTRB(700.0, 0.0, 1000.0, 600.0));
+
+    await tester.drag(find.byType(Carousel), const Offset(-50, 0), kind: PointerDeviceKind.trackpad);
+    await tester.pump();
+    // The item 0 should be pinned and has a size change from 350 to 50.
+    expect(tester.getRect(getItem(0)), const Rect.fromLTRB(0.0, 0.0, 300.0, 600.0));
+    // Keep dragging to left, extent of item 0 won't change (still 300) and part of item 0 will
+    // be off screen.
+    await tester.drag(find.byType(Carousel), const Offset(-50, 0), kind: PointerDeviceKind.trackpad);
+    await tester.pump();
+    expect(tester.getRect(getItem(0)), const Rect.fromLTRB(-50, 0.0, 250, 600));
+  });
+}
+
+Finder getItem(int index) {
+  return find.descendant(of: find.byType(Carousel), matching: find.ancestor(of: find.text('Item $index'), matching: find.byType(Padding)));
+}
+
+Future<TestGesture> _pointGestureToCarouselItem(WidgetTester tester, GlobalKey key) async {
```

Comment:
Maybe a better name:

```suggestion
Future<TestGesture> hoverPointerOverCarouselItem(WidgetTester tester, GlobalKey key) async {
```